wallet/src/Pos/Wallet/Web/Pending/Worker.hs

```diff
@@ -0,0 +1,167 @@
+-- | What should happen with existing pending transaction if attempt
+-- to resubmit it failed.
+processPtxFailure :: MonadPendings m => PendingTx -> ToilVerFailure -> m ()
+processPtxFailure PendingTx{..} e =
+    unless (isReclaimableFailure e) $ do
+        let newCond = PtxWontApply (sformat build e)
+        void $ casPtxCondition _ptxWallet _ptxTxId PtxApplying newCond
+        logInfo $ sformat ("Pending transaction "%build%" was canceled")
+                  _ptxTxId
+
+processPtxInUpperBlocks :: MonadPendings m => SlotId -> PendingTx -> m ()
+processPtxInUpperBlocks curSlot PendingTx{..} = do
+    mdepth <- getWalletAssuredDepth _ptxWallet
+    if | PtxInUpperBlocks slotId <- _ptxCond,
+         Just depth <- mdepth,
+         longAgo depth slotId -> do
+             void $ casPtxCondition _ptxWallet _ptxTxId _ptxCond PtxPersisted
+             logInfo $ sformat ("Transaction "%build%" got persistent") _ptxTxId
+       | otherwise -> pass
+  where
+     longAgo depth (flattenSlotId -> ptxSlot) =
+         ptxSlot + getBlockCount depth < flattenSlotId curSlot
+
+-- | 'True' for slots which are equal to
+-- @ptxCreationSlot + initialDelay + furtherDelay * k@ for some integer @k@
+whetherCheckPtxOnSlot :: HasCoreConstants => SlotId -> PendingTx -> Bool
+whetherCheckPtxOnSlot (flattenSlotId -> curSlot) ptx = do
+    let ptxSlot = flattenSlotId (_ptxCreationSlot ptx)
+        checkStartSlot = ptxSlot + initialDelay
+    and [ curSlot > checkStartSlot
+        , ((curSlot - checkStartSlot) `mod` furtherDelay) == 0
+        ]
+  where
+    initialDelay = 3 :: FlatSlotId
+    furtherDelay = 1 :: FlatSlotId
+
+resubmitTx :: MonadPendings m => SendActions m -> PendingTx -> m ()
+resubmitTx SendActions{..} ptx@PendingTx{..} = do
+    logInfo $ sformat ("Resubmitting tx "%build) _ptxTxId
+    -- FIXME [CSM-256] Doesn't it introduce a race condition?
+    void (submitAndSaveTx enqueueMsg _ptxTxAux) `catches` handlers
+  where
+    handlers =
+        [ Handler $ \e -> do
+            reportFail "Failed to resubmit tx" e
+            processPtxFailure ptx e
+
+        , Handler $ \(SomeException e) ->
+            reportFail "Failed to resubmit tx, will try again later" e
+        ]
+    reportFail :: MonadPendings m => Exception e => Text -> e -> m ()
+    reportFail desc =
+        logInfo . sformat (stext%" "%build%": "%shown) desc _ptxTxId
+
+-- | Distributes pending txs submition over current slot ~evenly
+resubmitPtxsDuringSlot
+    :: MonadPendings m
+    => SendActions m -> [PendingTx] -> m ()
+resubmitPtxsDuringSlot sendActions ptxs = do
+    interval <- evalSubmitDelay (length ptxs)
+    forM_ ptxs $ \ptx -> do
+        delay interval
+        fork $ resubmitTx sendActions ptx
+  where
+    submitionEta = 5 :: Second
+    evalSubmitDelay toResubmitNum = do
+        slotDuration <- getLastKnownSlotDuration
+        let checkPeriod = max 0 $ slotDuration - convertUnit submitionEta
+        return (checkPeriod `div` fromIntegral toResubmitNum)
```

Comment:
If `checkPeriod` < `toResubmitNum`, then I have to send everything immediately, that's ok.
`min toResubmitNum checkPeriod` doesn't make sense because one is measured in seconds, while another has no measure